DevTools/IPCCompiler/main.cpp

```diff
@@ -267,31 +282,58 @@ int main(int argc, char** argv)
         };
 
         auto do_message = [&](const String& name, const Vector<Parameter>& parameters, const String& response_type = {}) {
-            out() << "class " << name << " final : public IPC::Message {";
-            out() << "public:";
+            SourceGenerator message_generator { endpoint_generator };
+
+            message_generator.set("message.name", name);
+            message_generator.set("message.response_type", response_type);
+            message_generator.set("message.constructor", constructor_for_message(name, parameters));
+
+            message_generator.append(R"~~~(
+class @message.name@ final : public IPC::Message {
+public:
+)~~~");
+
             if (!response_type.is_null())
-                out() << "    typedef class " << response_type << " ResponseType;";
-            out() << "    " << constructor_for_message(name, parameters);
-            out() << "    virtual ~" << name << "() override {}";
-            out() << "    virtual i32 endpoint_magic() const override { return " << endpoint.magic << "; }";
-            out() << "    virtual i32 message_id() const override { return (int)MessageID::" << name << "; }";
-            out() << "    static i32 static_message_id() { return (int)MessageID::" << name << "; }";
-            out() << "    virtual const char* message_name() const override { return \"" << endpoint.name << "::" << name << "\"; }";
-            out() << "    static OwnPtr<" << name << "> decode(InputMemoryStream& stream, size_t& size_in_bytes)";
-            out() << "    {";
-
-            out() << "        IPC::Decoder decoder(stream);";
+                message_generator.append(R"~~~(
+   typedef class @message.response_type@ ResponseType;
+)~~~");
```

Comment:
Is it just me, or does this somehow look worse?
I kinda like the `@foobar.blahblah@`, but the use of raw strings and weird indents makes it look...odd

perhaps we could instead do something like `source_generator << "typedef class @message.response_type@ ResponseType";`, and have it magically indent for us?

Something that comes to mind is `source_generator.indented(1) << "blahblah"`
and newlines could be inserted implicitly between `<<`'s now, as they cannot meaningfully delimit anything else given this `@foo@` thing?

Reply:
Another option I can think of is what I did with the shell formatter [here](https://github.com/SerenityOS/serenity/blob/master/Shell/Formatter.cpp#L253-L256) and [here](https://github.com/SerenityOS/serenity/blob/ac8fe3d0625ffc95f9b042e990b122ccf5abe2d3/Shell/Formatter.cpp#L76-L79)
But that assumes stuff about the semantics of the language it's generating, so probably not a great "source generator" interface

Reply:
> Is it just me, or does this somehow look worse?

It's surely not just you, but I actually like it. Previously, it was difficult to "see" the file that is being generated, now it looks like two "interleaved" source files.

(I notice now, that I was a bit inconsistent with this, but more towards the end of the file I always used the multi-line string literals and did the indentation correctly. I think that really benefits helps see the generated file at a glance.)

Reply:
*(Now about the previous message.)*

I think if we did code generation on a much larger scale, then a more "functional" approach would be a really good idea, this is really just me trying to be create a simple solution.

Reply:
That's fair, I suppose.
it looks odd to me, but I guess I could get used to it :shrug:

Reply:
Another way of doing this would be like Go's text template thing. ([Go Documentation][1])

It's like `std::format` but with loops and conditions. For the simple stuff we are doing here that would totally suffice, but implementing something like that would really be overkill.

  [1]: https://golang.org/pkg/text/template/#hdr-Actions

Reply:
Now that's definitely overkill.
I don't think there's any reason to go down that route, also given that it'll end up looking like one big string without good syntax highlighting... `(include python/fstring-syntax-fight)`

Reply:
I might do some variation of

~~~c++
source_generator.indent(4) << "int foo = @foo@";
~~~

I'll have to sleep over that.

Reply:
I don't think the `indent(4)` stuff is really needed, since the compiler doesn't care about it and we can always run it through clang format. (The indentation only truly aids readability if it is present in the source code as-is.)

I could go with the `generator << "int foo = @foo@;` syntax though, take the following snippet for example:

~~~c++
        endpoint_generator.append(R"~~~(
            return nullptr;
        }
        i32 message_id = 0;
        stream >> message_id;
        if (stream.handle_any_error()) {
)~~~");
~~~

I could change this to one of the following snippets:

~~~c++
        endpoint_generator << R"~~~(
            return nullptr;
        }
        i32 message_id = 0;
        stream >> message_id;
        if (stream.handle_any_error()) {
)~~~";
~~~

~~~c++
        endpoint_generator << "        return nullptr";
        endpoint_generator << "        }";
        endpoint_generator << "        i32 message_id = 0;"
        endpoint_generator << "        stream >> message_id;"
        endpoint_generator << "        if (stream.handle_any_error()) {"
~~~

I don't think that improves the code a lot, although in cases where just one line is printed it might. (But for consistency I would all format them the same way.)

Reply:
What I thought `<<` could indicate was a logical statement, so newlines and indentation would be automatically inserted.
But that makes the generated source much less visible, so maybe your raw strings _are_ the same approach

Reply:
I'll just leave it as it is now, but I'll keep your thoughts in the back of my head, I usually come back to this sort of stuff.

I think we are doing some code generation in LibWeb too, so when I go and eradicate the printf and LogStream stuff from there, maybe a more "abstract" approach would be better there?